spec/models/user_spec.rb

```diff
@@ -1,5 +1,73 @@
-require 'rails_helper'
+require "rails_helper"
 
 RSpec.describe User, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe "enum for role" do
+    it {is_expected.to define_enum_for(:role).with_values(admin: 0, customer: 1)}
+  end
+
+  describe "associations" do
+    it {is_expected.to have_many(:orders).dependent(:destroy)}
+    it {is_expected.to have_many(:comments).dependent(:destroy)}
+  end
+
+  describe "validates" do
+    context "validate presence" do
+      it {is_expected.to validate_presence_of :email}
+      it {is_expected.to validate_presence_of :name}
+      it {is_expected.to validate_presence_of :password}
+    end
+
+    context "validate length of name" do
+      it {is_expected.to validate_length_of(:name).is_at_least(Settings.user.name_validates.name_min_length).is_at_most(Settings.user.name_validates.name_max_length)}
+    end
+
+    context "validate length of password" do
+      it {is_expected.to validate_length_of(:password).is_at_least(Settings.user.password_validates.password_min_length)}
+    end
+
+    context "validate length of address" do
+      it {is_expected.to validate_length_of(:address).is_at_least(Settings.user.address_min_length).is_at_most(Settings.user.address_max_length)}
+    end
+
+    context "validate length of phone" do
+      it {is_expected.to validate_length_of(:phone_num).is_at_least(Settings.user.min_phone).is_at_most(Settings.user.max_phone)}
+    end
+  end
+
+  describe "scope" do
+    let!(:user1) {FactoryBot.create :user, name: "Tien", role: 0}
```

Comment:
`let` vs `let!` khác nhau ntn?

Reply:
khai báo let chỉ có giá trị trong từng example (trong Rspec thì được đóng gói trong khối it) và nếu như bạn gọi đến một biến let nhiều lần trong cùng một example thì nó sẽ chỉ khai báo khối let một lần duy nhất. let! được gọi trong một khối before ngầm, do đó kết quả đã được tính toán và cache lại trước khi khổi it được thực thi.